independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java

```diff
@@ -0,0 +1,158 @@
+import javax.inject.Singleton;
+
+/**
+ * Encapsulates all contexts used in the container.
+ * <p>
+ * The idea is to avoid unnecessary map usage for built-in contexts.
+ */
+class Contexts {
+
+    // Built-in contexts
+    final ManagedContext requestContext;
+    final InjectableContext applicationContext;
+    final InjectableContext singletonContext;
+
+    // Used to optimize the getContexts(Class<? extends Annotation>)
+    private final List<InjectableContext> applicationContextSingleton;
+    private final List<InjectableContext> singletonContextSingleton;
+    private final List<InjectableContext> requestContextSingleton;
+
+    // Lazily computed list of contexts for a scope
+    private final ClassValue<List<InjectableContext>> unoptimizedContexts;
+
+    // Precomputed values
+    final int size;
+    final Set<Class<? extends Annotation>> scopes;
+
+    Contexts(ManagedContext requestContext, InjectableContext applicationContext,
+            InjectableContext singletonContext, Map<Class<? extends Annotation>, List<InjectableContext>> contexts) {
+        this.requestContext = requestContext;
+        this.applicationContext = applicationContext;
+        this.singletonContext = singletonContext;
+
+        this.applicationContextSingleton = Collections.singletonList(applicationContext);
+        this.singletonContextSingleton = Collections.singletonList(singletonContext);
+        // request singleton is only set if there is only one context registered for @RequestScoped (99% of cases)
+        List<InjectableContext> requestContexts = contexts.get(RequestScoped.class);
+        this.requestContextSingleton = requestContexts != null ? requestContexts : Collections.singletonList(requestContext);
+
+        if (!contexts.isEmpty()) {
+            // At least one custom context is registered
+            int mapSize = contexts.size() + 1;
+            @SuppressWarnings("unchecked")
+            Class<? extends Annotation>[] keys = new Class[mapSize];
+            @SuppressWarnings("unchecked")
+            List<InjectableContext>[] values = new List[mapSize];
+            int index = 0;
+            for (Map.Entry<Class<? extends Annotation>, List<InjectableContext>> entry : contexts.entrySet()) {
+                keys[index] = entry.getKey();
+                values[index] = entry.getValue();
+                index++;
+            }
+            this.unoptimizedContexts = new ClassValue<>() {
+                @Override
+                protected List<InjectableContext> computeValue(Class<?> type) {
+                    for (int i = 0; i < mapSize; i++) {
+                        if (keys[i].equals(type)) {
+                            return values[i];
+                        }
+                    }
+                    return Collections.emptyList();
+                }
+            };
+            this.size = 3 + mapSize;
+            Set<Class<? extends Annotation>> all = new HashSet<>(contexts.keySet());
```

Comment:
If `contexts` are not empty, but do _not_ contain a custom implementation of the request context, then `all` won't include `RequestScoped.class` I think?

Reply:
You're right, that's an oversight... it was there a few iterations earlier.